components/style/values/specified/effects.rs

```diff
@@ -0,0 +1,247 @@
+
+//! Specified types for CSS values related to effects.
+
+use cssparser::{BasicParseError, Parser, Token};
+use parser::{Parse, ParserContext};
+#[cfg(feature = "gecko")]
+use std::fmt;
+use style_traits::ParseError;
+#[cfg(not(feature = "gecko"))]
+use style_traits::StyleParseError;
+#[cfg(feature = "gecko")]
+use style_traits::ToCss;
+use values::computed::{Context, Number as ComputedNumber, ToComputedValue};
+use values::computed::effects::DropShadow as ComputedDropShadow;
+use values::generics::effects::Filter as GenericFilter;
+use values::generics::effects::FilterList as GenericFilterList;
+use values::specified::{Angle, Percentage};
+#[cfg(feature = "gecko")]
+use values::specified::color::Color;
+use values::specified::length::Length;
+#[cfg(feature = "gecko")]
+use values::specified::url::SpecifiedUrl;
+
+/// A specified value for the `filter` property.
+pub type FilterList = GenericFilterList<Filter>;
+
+/// A specified value for a single `filter`.
+pub type Filter = GenericFilter<Angle, Factor, Length, DropShadow>;
+
+/// A value for the `<factor>` parts in `Filter`.
+///
+/// Should be `NumberOrPercentage`, but Gecko doesn't support that yet.
+#[cfg_attr(feature = "servo", derive(Deserialize, HeapSizeOf, Serialize))]
+#[derive(Clone, Debug, HasViewportPercentage, PartialEq, ToCss)]
+pub enum Factor {
+    /// Literal number.
+    Number(ComputedNumber),
+    /// Literal percentage.
+    Percentage(Percentage),
+}
+
+/// A specified value for the `drop-shadow()` filter.
+///
+/// Currently unsupported outside of Gecko.
+#[cfg(not(feature = "gecko"))]
+#[cfg_attr(feature = "servo", derive(Deserialize, HeapSizeOf, Serialize))]
+#[derive(Clone, Debug, HasViewportPercentage, PartialEq, ToCss)]
+pub enum DropShadow {}
+
+/// A specified value for the `drop-shadow()` filter.
+///
+/// Contrary to the canonical order from the spec, the color is serialised
+/// first, like in Gecko's computed values and in all Webkit's values.
+#[cfg(feature = "gecko")]
+#[derive(Clone, Debug, HasViewportPercentage, PartialEq)]
+pub struct DropShadow {
+    /// Color.
+    pub color: Option<Color>,
+    /// Horizontal radius.
+    pub horizontal: Length,
+    /// Vertical radius.
+    pub vertical: Length,
+    /// Blur radius.
+    pub blur: Option<Length>,
+}
+
+impl Parse for FilterList {
+    #[inline]
+    fn parse<'i, 't>(
+        context: &ParserContext,
+        input: &mut Parser<'i, 't>
+    ) -> Result<Self, ParseError<'i>> {
+        let mut filters = vec![];
+        while let Ok(filter) = input.try(|i| Filter::parse(context, i)) {
+            filters.push(filter);
+        }
+        if filters.is_empty() {
+            input.expect_ident_matching("none")?;
+        }
+        Ok(GenericFilterList(filters.into_boxed_slice()))
+    }
+}
+
+impl Parse for Filter {
+    #[inline]
+    fn parse<'i, 't>(
+        context: &ParserContext,
+        input: &mut Parser<'i, 't>
+    ) -> Result<Self, ParseError<'i>> {
+        match true {
```

Comment:
This is not stable on 1.15 on statements. Feel free to complain to the Gecko people for us to get a stable bump. :)